lib/associations/mixin.js

```diff
@@ -90,13 +105,20 @@ function singleLinked(Type) {
     options.hooks = options.hooks === undefined ? false : Boolean(options.hooks);
     options.useHooks = options.hooks;
 
+    if (options.useHooks) {
+      this.runHooks('beforeAssociate', source, target, options);
+    }
     // the id is in the foreign table
     const association = new Type(source, target, _.extend(options, source.options));
     source.associations[association.associationAccessor] = association;
 
     association._injectAttributes();
     association.mixin(source.prototype);
 
+    if (options.useHooks) {
+      this.runHooks('afterAssociate', target, options);
```

Comment:
Instead of `target`, pass `association` (that is what I meany by emit association :p)

Reply:
No, we just want to know what association got created, source and target info is available on this object

Reply:
@sushantdhiman Ok. Here is another stab at it. I'm not passing source or target anymore since they are available within hook through `this`. We're passing association object now. I'm open to more suggestions. Tests should be passing but they might be a bit wonky so I'll accept any tips.